### PR TITLE
Confidence intervals

### PR DESCRIPTION
**Pull request by cpennington**

### CODE REVIEW

**Comment by cpennington** [comment]
@doctoryes @cgoldberg @jimabramson @BenjiLee: Confidence intervals!

**Comment by cpennington** [comment]
Depends on #4

File: locust/stats.py [hunk 1]
```diff
@@ -63,63 +79,145 @@ def clear_all(self):
...
+
+    def confidence_interval_column_name(self, percentile):
+        """
+        Return a ErrorBar of column names for the (lower, upper) error bar
+        values for the `percentile` value.
+        """
+        return ErrorBar(
+            "{:.0%} low error bound".format(percentile),
+            "{:.0%} high error bound".format(percentile),
+        )
```
Comment (doctoryes) [inline locust/stats.py]: Both bounds are the same here?

File: locust/stats.py [hunk 2]
```diff
@@ -63,63 +79,145 @@ def clear_all(self):
...
+        for percentile in PERCENTILES:
+            data.headers.append(self.percentile_column_name(percentile))
+
+            if CONFIDENCE_INTERVALS:
+                data.headers.extend(self.confidence_interval_column_name(percentile))
+
+        data.headers.append("100%")
+
+        for _, stats in sorted(self.entries.iteritems()):
+            data.append(stats.percentile(include_empty))
```
Comment (doctoryes) [inline locust/stats.py]: `itervalues()`?

File: locust/stats.py [hunk 3]
```diff
@@ -337,24 +438,54 @@ def get_response_time_percentile(self, percent):
...
-            self.get_response_time_percentile(0.99),
-            self.max_response_time
+    @property
+    def jittered_response_times(self):
+        """
+        A numpy array containing all of the response times for this StatsEntry, jittered
+        by a random amount between 0 and 0.001.
+
+        This prevents the bootstrap process from falling over in the jacknife step
+        if all of the jacknifed stats are equal.
```
Comment (doctoryes) [inline locust/stats.py]: Perhaps you should make clear where the error lies that makes it fall over? It's not clear whether this PR introduces an instability or if it already existed.

**Comment by doctoryes** [comment]
Only found nits - :+1:

File: setup.py [hunk 4]
```diff
@@ -44,12 +44,15 @@ def run(self):
...
-    install_requires=["gevent>=1.0.1", "flask>=0.10.1", "requests>=2.4.1", "msgpack-python>=0.4.2"],
+    install_requires=["gevent>=1.0.1", "flask>=0.10.1", "requests>=2.4.1", "msgpack-python>=0.4.2", "tablib", "tabulate"],
     tests_require=['unittest2', 'mock', 'pyzmq'],
     entry_points={
         'console_scripts': [
             'locust = locust.main:main',
         ]
     },
+    extras_require={
+        'scipy': ['scipy', 'scikits.bootstrap'],
```
Comment (cgoldberg) [inline setup.py]: needs numpy also?

**Comment by cgoldberg** [comment]
it might be easier to get this merged upstream if you didn't touch the whitespace/newlines (to keep scope down)  Instead, fix those in a separate PR.

btw, locust doesn't follow PEP8 very well.. It has 1002 violations across the codebase.  I'm tempted to run autopep8 and fix all of them in one shot.... but I'm not sure if the maintainers would find such a PR helpful or just annoying...  I suppose I should try.

File: locust/stats.py (hunk 1 above)
Comment (cpennington) [inline locust/stats.py]: The names of the columns follow the same pattern.

File: locust/stats.py [hunk 5]
```diff
@@ -449,40 +585,41 @@ def on_slave_report(client_id, data):
...
-
-    console_logger.info((" %-" + str(STATS_NAME_WIDTH) + "s %7d %12s %42.2f") % ('Total', total_reqs, "%d(%.2f%%)" % (total_failures, fail_percent), total_rps))
+    data = stats.get_request_stats_dataset()
+    console_logger.info(tabulate(data.dict, headers="keys"))
     console_logger.info("")
 
 def print_percentile_stats(stats):
+    data = stats.get_percentile_dataset()
+
+    if CONFIDENCE_INTERVALS:
```
Comment (jimabramson) [inline locust/stats.py]: suggestion: print out a message to the console advising patience while these computations are being done.   I just did ctrl+c on a pretty small test run and the console hung for 5 seconds before i saw any output.

Comment (cpennington) [inline locust/stats.py]: It should be significantly faster now, but yes, I'll add a logging message.

**Pull request merged**